### PR TITLE
refactor: rewrite  framework with good UI

### PR DESCRIPTION
**Pull request by ije**
- brand-new UI 
- use https://deno.land/x/htm instead of nano_jsx
- support customized header/footer with JSX
- support external links with icons

Screen Shot 2022-06-02 at 18 39 31

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (bartlomieju) [inline .vscode/settings.json:1]: Remove

File: README.md [hunk 2]
```diff
@@ -33,15 +34,19 @@ from the Deno Manual.
 
 You can customize your blog as follows:
 
-```js
+```jsx
```
Comment (bartlomieju) [inline README.md:37]: Why jsx?

File: blog_test.tsx [hunk 3]
```diff
@@ -1,21 +1,21 @@
 // Copyright 2022 the Deno authors. All rights reserved. MIT license.
 
-import { configureBlog, createBlogHandler, redirects } from "./blog.tsx";
+/** @jsx h */
```
Comment (bartlomieju) [inline blog_test.tsx:3]: Why is it needed?

File: testdata/my_blog.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/** @jsx h */
+
```
Comment (bartlomieju) [inline testdata/my_blog.tsx:2]: Seems like this file is not used in tests?

File: middlewares.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import { createReporter } from "./deps.ts";
+import type { BlogContext, BlogMiddleware } from "./types.d.ts";
+
+export function ga(gaKey: string): BlogMiddleware {
+  if (gaKey.length === 0) {
+    throw new Error("GA key cannot be empty.");
```
Comment (bartlomieju) [inline middlewares.ts:6]: I think it's superfluous to move it to a different file

File: hmr.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 the Deno authors. All rights reserved. MIT license.
+
+const HMR_CLIENT = `let socket;
```
Comment (bartlomieju) [inline hmr.ts:3]: I think it's superfluous to move it to a separate file at this point

File: components.tsx [hunk 7]
```diff
@@ -6,137 +6,309 @@
...
+        </a>
+      </p>
     </div>
   );
 }
 
-export function PostPage(
-  { post, hmr, state }: { post: Post; hmr: boolean; state: BlogState },
-) {
+type PostPageProps = {
```
Comment (bartlomieju) [inline components.tsx:131]: IIRC it's preferable to use `interface` instead of `type` for "objects"

File: blog.tsx [hunk 8]
```diff
@@ -437,3 +325,6 @@ function serveRSS(
     },
   });
 }
+
+export * from "https://deno.land/x/htm@0.0.6/mod.tsx";
```
Comment (bartlomieju) [inline blog.tsx:329]: This should be added to `deps.ts` and reexported from there, but why does it need to be reexported in the first place?

File: blog.tsx [hunk 9]
```diff
@@ -332,62 +251,90 @@ export async function handler(
...
+  let fsRoot = blogState.directory;
+  try {
+    await Deno.lstat(join(blogState.directory, "./posts", pathname));
+    fsRoot = join(blogState.directory, "./posts");
+  } catch (e) {
+    if (!(e instanceof Deno.errors.NotFound)) {
+      console.error(e);
+      return new Response(e.message, { status: 500 });
+    }
   }
```
Comment (bartlomieju) [inline blog.tsx:335]: Why change this bit? It worked perfectly fine

File: blog.tsx [hunk 10]
```diff
@@ -437,3 +384,64 @@ function serveRSS(
...
+          "location": maybeRedirect,
+        },
+      });
+    }
+
+    return await ctx.next();
+  };
+}
+
+export { Fragment, h };
```
Comment (bartlomieju) [inline blog.tsx:447]: Why is this export needed?

Comment (ije) [reply blog.tsx:447]: to support customized header/footer with jsx

```ts
Blog({
  header: <header>...</header>,
  footer: <footer>...</footer>
})
```

Comment (bartlomieju) [reply blog.tsx:447]: I see, makes sense; a small nitpick: can you move it to the top?

File: blog.tsx (hunk 9 above)
Comment (bartlomieju) [reply blog.tsx:335]: Okay sure

File: blog.tsx (hunk 10 above)
Comment (ije) [reply blog.tsx]: moved

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**